### PR TITLE
Accept compact MXFP8 scales in GEMM preparation

### LINKED ISSUE #3518: MXFP8 GEMM rejects compact unswizzled scales that quantize_ accepts
For an MXFP8Tensor with `with_gemm_swizzled_scales=False`, `quantize_()` accepts compact scale buffers, but `general_gemm()` requires the caller to pad those buffers to GEMM tile dimensions before TE prepares them for GEMM.

For a 64×64 tensor, the logical scale shapes are `(64, 2)` rowwise and `(2, 64)` columnwise. Quantization succeeds with these actual allocations, preserves their shapes and storage, and produces the same logical data and scales as an independently quantized TE reference. GEMM then rejects them, expecting `(128, 4)` and `(4, 128)`, respectively. Using those padded allocations succeeds with the same data shape.

Expected behavior: GEMM should accept compact unswizzled scales and handle padding internally, either before swizzling or by reading the compact input directly into a padded, swizzled output. The GEMM-ready representation can retain its tile constraints without requiring the unswizzled storage representation to carry padding.

Reproducer (single Blackwell GPU; save as repro.py):

```python
import sys

import torch
import transformer_engine_torch as tex
from transformer_engine.pytorch.cpp_extensions import general_gemm
from transformer_engine.pytorch.tensor.mxfp8_tensor import MXFP8Quantizer, MXFP8Tensor

layout = sys.argv[1]  # TN or NN
padded = "--padded" in sys.argv[2:]
quantizer = MXFP8Quantizer(tex.DType.kFloat8E4M3)

def empty(shape):
    return torch.empty(shape, dtype=torch.uint8, device="cuda")

weight = MXFP8Tensor(
    shape=(64, 64),
    dtype=torch.bfloat16,
    device="cuda",
    rowwise_data=empty((64, 64)),
    columnwise_data=empty((64, 64)),
    rowwise_scale_inv=empty((128, 4) if padded else (64, 2)),
    columnwise_scale_inv=empty((4, 128) if padded else (2, 64)),
    fp8_dtype=tex.DType.kFloat8E4M3,
    quantizer=quantizer,
    with_gemm_swizzled_scales=False,
)
weight.quantize_(torch.randn((64, 64), dtype=torch.float32, device="cuda"))
torch.cuda.synchronize()
print("quantize_ succeeded", flush=True)

activation = quantizer(torch.randn((64, 64), dtype=torch.float32, device="cuda"))
general_gemm(weight, activation, out_dtype=torch.bfloat16, layout=layout)
torch.cuda.synchronize()
print("GEMM succeeded", flush=True)
```

Run each case in a separate process:

```bash
python repro.py TN           # Quantization succeeds; GEMM rejects rowwise scales.
python repro.py NN           # Quantization succeeds; GEMM rejects columnwise scales.
python repro.py TN --padded  # Both succeed.
python repro.py NN --padded  # Both succeed.
```

Observed errors (relevant excerpts):

```text
TN: transformer_engine/common/transformer_engine.cpp:103 in function CheckScaleTensorShape:
invalid scale_inv shape (expected (128,4), got (64,2))

NN: transformer_engine/common/transformer_engine.cpp:112 in function CheckScaleTensorShape:
invalid columnwise_scale_inv shape (expected (4,128), got (2,64))
```

Environment of the reproduced failure:

- GPU: NVIDIA GB200
- Transformer Engine: 2.18.0+27486e03
- PyTorch: 2.14.0a0+4fdf77b940.nv26.08
- CUDA reported by PyTorch: 13.4

Motivation: Megatron FSDP v2 stores and all-gathers compact, unswizzled MXFP8 planes, for memory. Requiring callers to pad scales before GEMM adds allocations and copies in the wrapper. TE already prepares unswizzled scales for GEMM, so it would be useful for that preparation to accept the compact representation directly.

### PR DESCRIPTION
**Pull request by wujingyue**
`general_gemm()` rejects compact MXFP8 scales even when `with_gemm_swizzled_scales=False`: a 64×64 tensor requires rowwise scales of `(128, 4)` instead of `(64, 2)`, or columnwise scales of `(4, 128)` instead of `(2, 64)`.

Accept exact compact MXFP8 scale shapes in the single-tensor swizzle API and read them directly into a padded, swizzled output. A compact-input CUDA kernel uses independent input strides and fills padding with zeros, without an intermediate padding allocation or copy. The PyTorch helper allocates the required output shape. Already-padded inputs retain the existing kernel; GEMM preparation preserves the caller's buffers and unswizzled flag.

Fixes #3518 for the PyTorch single-GEMM path. Grouped swizzle is unchanged.

Validation:
- Built matching TE common and PyTorch libraries from this checkout on NVIDIA GB200.
- All 30 focused tests passed: 18 GEMM cases covering both operands, TN/NN/NT, square/rectangular shapes, compact/padded storage, and input preservation; 12 byte-layout cases covering both orientations and padding contents.
- Compute Sanitizer memcheck: all 12 byte-layout cases passed with zero errors.
- Captured GPU microbenchmarks measured 1.1–4.3 µs for direct swizzling versus 2.9–11.1 µs for padding followed by the existing swizzle, across shapes 64×64, 96×160, 4096×4128, and 16384×4128 in both orientations (about 2.5–4.2× faster). These are preparation timings, not end-to-end training speedups.
- cpplint, isort, Black, clang-format, and whitespace checks passed. Full TE L0 suites have not been run.

The GEMM tests construct compact inputs from independently quantized reference planes to isolate swizzle/GEMM behavior. Compact in-place quantization showed separate rectangular numerical mismatches during investigation; this patch does not address those.

### CODE REVIEW

**Review by vthumbe1503 — COMMENTED** [review]
@wujingyue Thanks for the PR!

I wanted to understand what is the use-case where we have unpadded scales in the mxfp8 tensor?

TE quantizer always produces padded scales in the MXFP8 Tensor. And so we should never face this issue where the scales are not padded to (128,4) or (4,128). If external framework is constructing/manipulating the MXFP8Tensor, then I would say, it is the responsibility of that framework and not TE, to pad the scales tensor.

If the external framework cannot sacrifice the performance of additional padding. We can potentially fuse the padding within the swizzling kernel in a way swizzling kernel accepts unpadded scales and fills zero manually in those unpadded locations. We have done something similar for the grouped swizzling kernel.

**Comment by wujingyue** [comment]
> I wanted to understand what is the use-case where we have unpadded scales in the mxfp8 tensor?

Sure. Have you seen the issue this PR links to? I've been adding context there because this PR is just for demo at this moment. Shall we follow up over there? 

> TE quantizer always produces padded scales in the MXFP8 Tensor. 

Not really -- according to #3518, quantizer itself works great with unpadded scales. But I could certainly be missing something! 

cc @vthumbe1503

**Comment by vthumbe1503** [comment]
In the issue you mentioned, custom MXFP8Tensor is being created for weight like this.

```
weight = MXFP8Tensor(
    shape=(64, 64),
    dtype=torch.bfloat16,
    device="cuda",
    rowwise_data=empty((64, 64)),
    columnwise_data=empty((64, 64)),
    rowwise_scale_inv=empty((128, 4) if padded else (64, 2)),
    columnwise_scale_inv=empty((4, 128) if padded else (2, 64)),
    fp8_dtype=tex.DType.kFloat8E4M3,
    quantizer=quantizer,
    with_gemm_swizzled_scales=False,
)
```

I see that this is used in Megatron FSDP v2. So I see that you are dealing with weight sharding in FSDP which can be tricky since sharded tensors might not be padded? 


> quantizer itself works great with unpadded scales

I understand quantization works with unpadded scales. So the problem seems to be that GEMMS need swizzled scales padded. 

So 2 solutions can be
1. Pad the compact scales in megatron itself before constructing the MXFP8Tensor shard for the weights. We do something similar in FSD2P. https://github.com/NVIDIA/TransformerEngine/blob/796346c0e0497b1f56a8d36ec76e02db5a6fed47/transformer_engine/pytorch/tensor/mxfp8_tensor.py#L718
2. Or instead of padding and creating a copy for the compact scales.  How about we handle padding directly in the swizzling kernel, to avoid padding performance penalty? We do something similar in grouped swizzling kernel https://github.com/NVIDIA/TransformerEngine/blob/796346c0e0497b1f56a8d36ec76e02db5a6fed47/transformer_engine/common/swizzle/swizzle.cu#L376-L377. Essentially swizzling kernel will be aware that the scales are not padded already. So manually fills them with zeros.


With the current solution, we are silently introducing a performance penalty in the framework by doing the padding, which I am hoping to avoid.

**Comment by wujingyue** [comment]
> How about we handle padding directly in the swizzling kernel, to avoid padding performance penalty

I asked my agent to one-shot it. I really don't understand what it's doing. Could you or someone from the TE team take over? @vthumbe1503
